### PR TITLE
Expose choice of symbology reference scale for vector layers

### PR DESCRIPTION
**Pull request by nyalldawson**
If set, the reference scale indicates the map scale at which symbology and labeling sizes which uses paper-based units
(such as millimeters or points) relate to. The sizes will be scaled accordingly whenever the map is viewed at a different
scale.

For instance, a line layer using a 2mm wide line with a 1:2000 reference scale set will be rendered using 4mm wide lines when
the map is viewed at 1:1000.

Sponsored by North Road, thanks to SLYR

### CODE REVIEW

**Comment by andreasneumann** [comment]
Hi @nyalldawson ,

Will the maximum barriers still apply to avoid for symbols ( https://www.qgis.org/en/site/forusers/visualchangelog316/index.html#layout-legend-maximum-marker-size )?

I think it is important to avoid legends with map units displaying too wide/large.

**Comment by nyalldawson** [comment]
@andreasneumann legends aren't affected by this -- even with a reference scale set a 2mm wide line will always show as 2mm wide in the legend

**Comment by nyalldawson** [comment]
Here's a few reasons why this approach is preferable to using mapunit sizes in some situations:

1. unless you are trying to represent a real-world distance, thinking in map units is strange for symbology. ie.. it's fine if i want to represent a road as a 10m wide line ... but setting font buffer size to "1.5m " is very counter-intuitive
2. If you are designing a layer for a particular map scale, it's annoying to do the maths to convert to map units... vs "i know it needs to be 10pt at 1:170000"
3. If you need to change this scale, good luck if you've used map units... you'll have to hunt through the layer and find ALL the different places you've set sizes using map units and change them. yuck.
4. If you want to make portable styles, map units are basically a no-go... they are too dependant on crs and other map properties

File: src/core/qgsrendercontext.cpp [hunk 1]
```diff
@@ -381,13 +383,18 @@ double QgsRenderContext::convertToPainterUnits( double size, QgsUnitTypes::Rende
       convertedSize = std::min( convertedSize, scale.maxSizeMM * mScaleFactor );
   }
 
+  const double symbologyReferenceScaleFactor = mSymbologyReferenceScale > 1 ? mSymbologyReferenceScale / mRendererScale : 1;
```
Comment (nirvn) [inline src/core/qgsrendercontext.cpp:386]: Shouldn't the conditional check be something like mSymbologyReferenceScale != -1 ?

File: src/core/symbology/qgsrenderer.h [hunk 2]
```diff
@@ -490,6 +527,17 @@ class CORE_EXPORT QgsFeatureRenderer
...
+    /**
+     * Clones generic renderer data to another renderer.
+     * Currently clones
+     *
+     * - Order By
+     * - Paint Effect
+     *
+     * \param destRenderer destination renderer for copied effect
+     */
+    void copyRendererData( QgsFeatureRenderer *destRenderer ) const;
```
Comment (nirvn) [inline src/core/symbology/qgsrenderer.h:539]: Nice move!

File: src/core/symbology/qgsrenderer.h [hunk 3]
```diff
@@ -490,6 +527,17 @@ class CORE_EXPORT QgsFeatureRenderer
...
 
+    /**
+     * Clones generic renderer data to another renderer.
+     * Currently clones
+     *
+     * - Order By
+     * - Paint Effect
+     *
+     * \param destRenderer destination renderer for copied effect
+     */
```
Comment (nirvn) [inline src/core/symbology/qgsrenderer.h:538]: \since

**Review by nirvn — APPROVED** [review]
Looks good, just a few comments.

File: src/core/qgsrendercontext.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsrendercontext.cpp:386]: Yep, fixed

**Pull request merged**